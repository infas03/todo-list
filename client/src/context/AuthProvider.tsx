import {
  createContext,
  useContext,
  useState,
  ReactNode,
  useEffect,
} from "react";
import { useDispatch, useSelector } from "react-redux";

import { Role } from "../constants/role";
import { User } from "../types";
import { RootState } from "../redux/store";
import {
  loginUserSuccess,
  logoutUserSuccess,
} from "../redux/actions/userAction";

type AuthContextType = {
  isLoggedIn: boolean;
  role: string | null;
  token: string | null;
  userDetails: User | null;
  resolved: boolean;
  login: (newToken: string, newRole: Role, userDetails: User) => void;
  logout: () => void;
};

const AuthContext = createContext<AuthContextType | undefined>(undefined);

type AuthProviderProps = {
  children: ReactNode;
};

export const AuthProvider = ({ children }: AuthProviderProps) => {
  const dispatch = useDispatch();

  const userDetailsState = useSelector(
    (state: RootState) => state.user?.userDetails,
  );

  const [isLoggedIn, setIsLoggedIn] = useState(false);
  const [role, setRole] = useState<string | null>(null);
  const [token, setToken] = useState<string | null>(
    localStorage.getItem("token"),
  );
  const [userDetails, setUserDetails] = useState<User | null>(null);
  const [resolved, setResolved] = useState(false);

  useEffect(() => {
    const storedToken = localStorage.getItem("token");

    if (storedToken && userDetailsState) {
      setIsLoggedIn(true);
      setToken(storedToken);
      setRole(userDetailsState.role);
      setUserDetails(userDetailsState ? userDetailsState : null);
    }

    setResolved(true);
  }, []);

  const login = (newToken: string, newRole: Role, newUserDetails: User) => {
    localStorage.setItem("token", newToken);
    dispatch(loginUserSuccess(newUserDetails));
    setToken(newToken);
    setRole(newRole);
    setUserDetails(newUserDetails);
    setIsLoggedIn(true);
  };

  const logout = () => {
    localStorage.removeItem("token");
    dispatch(logoutUserSuccess());
    setToken(null);
    setRole(null);
    setUserDetails(null);
    setIsLoggedIn(false);
  };

  return (
    <AuthContext.Provider
      value={{ isLoggedIn, token, userDetails, role, resolved, login, logout }}
    >
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = () => {
  const context = useContext(AuthContext);

  if (!context) {
    throw new Error("useAuth must be used within an AuthProvider");
  }

  return context;
};
