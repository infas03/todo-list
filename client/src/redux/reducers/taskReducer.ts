import { FETCH_TASKS } from "../actions/taskAction";
import { LOGOUT_SUCCESS } from "../actions/userAction";
import { TaskState } from "../types";

const initialState: TaskState = {
  tasks: [],
  totalTasks: 0,
  finishedTasks: 0,
};

const taskReducer = (state = initialState, action: any) => {
  switch (action.type) {
    case FETCH_TASKS:
      return {
        ...state,
        tasks: action.payload,
      };
    case LOGOUT_SUCCESS:
      return initialState;
    default:
      return state;
  }
};

export default taskReducer;
