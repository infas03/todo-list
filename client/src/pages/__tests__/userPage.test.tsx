import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";

import User from "../userPage";

import { AuthProvider } from "@/context/AuthProvider";
import taskReducer from "@/redux/reducers/taskReducer";

beforeEach(() => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: vi.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: vi.fn(),
      removeListener: vi.fn(),
      addEventListener: vi.fn(),
      removeEventListener: vi.fn(),
      dispatchEvent: vi.fn(),
    })),
  });
});

const mockStore = configureStore({
  reducer: {
    task: taskReducer,
  },
});

vi.mock("@/components/navbar", () => ({
  Navbar: () => <div>Mock Navbar</div>,
}));

describe("User Component", () => {
  it("renders the page title and subtitle", () => {
    render(
      <BrowserRouter>
        <Provider store={mockStore}>
          <AuthProvider>
            <User />
          </AuthProvider>
        </Provider>
      </BrowserRouter>,
    );

    expect(screen.getByTestId("page-title")).toHaveTextContent("My Tasks");
    expect(screen.getByTestId("page-subtitle")).toHaveTextContent(
      "Manage and track your assigned tasks",
    );
  });

  it("displays the alert with correct description", () => {
    render(
      <BrowserRouter>
        <Provider store={mockStore}>
          <AuthProvider>
            <User />
          </AuthProvider>
        </Provider>
      </BrowserRouter>,
    );

    const alert = screen.getByTestId("user-alert");

    expect(alert).toBeInTheDocument();
    expect(alert).toHaveTextContent(/sort your tasks by status/i);
  });

  it("renders the task table container", () => {
    render(
      <BrowserRouter>
        <Provider store={mockStore}>
          <AuthProvider>
            <User />
          </AuthProvider>
        </Provider>
      </BrowserRouter>,
    );

    expect(screen.getByTestId("task-table-container")).toBeInTheDocument();
  });
});
