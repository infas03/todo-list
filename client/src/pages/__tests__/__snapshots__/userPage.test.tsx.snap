// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserPage > should match the visual structure snapshot 1`] = `
<div>
  <div
    data-testid="default-layout"
  >
    <section
      class="flex flex-col items-center justify-center gap-4 py-8 md:py-0 "
      data-testid="user-page-section"
    >
      <div
        class="min-w-[900px]"
        data-testid="user-page-container"
      >
        <h1
          class="text-3xl font-bold"
          data-testid="page-title"
        >
          My Tasks
        </h1>
        <p
          class="text-lg text-gray-400"
          data-testid="page-subtitle"
        >
          Manage and track your assigned tasks
        </p>
        <div
          class="flex flex-grow flex-row w-full items-start py-3 px-4 gap-x-1 border-small rounded-medium bg-transparent border-primary text-primary mt-10"
          data-has-description="true"
          data-testid="user-alert"
          data-visible="true"
          role="alert"
        >
          <div
            class="flex-none relative w-9 h-9 rounded-full grid place-items-center shadow-small bg-primary-100 dark:bg-primary-50"
          >
            <svg
              class="fill-current w-6 absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2"
              fill="none"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 22C17.51 22 22 17.51 22 12C22 6.49 17.51 2 12 2C6.49 2 2 6.49 2 12C2 17.51 6.49 22 12 22ZM12.75 16C12.75 16.41 12.41 16.75 12 16.75C11.59 16.75 11.25 16.41 11.25 16L11.25 11C11.25 10.59 11.59 10.25 12 10.25C12.41 10.25 12.75 10.59 12.75 11L12.75 16ZM11.08 7.62C11.13 7.49 11.2 7.39 11.29 7.29C11.39 7.2 11.5 7.13 11.62 7.08C11.74 7.03 11.87 7 12 7C12.13 7 12.26 7.03 12.38 7.08C12.5 7.13 12.61 7.2 12.71 7.29C12.8 7.39 12.87 7.49 12.92 7.62C12.97 7.74 13 7.87 13 8C13 8.13 12.97 8.26 12.92 8.38C12.87 8.5 12.8 8.61 12.71 8.71C12.61 8.8 12.5 8.87 12.38 8.92C12.14 9.02 11.86 9.02 11.62 8.92C11.5 8.87 11.39 8.8 11.29 8.71C11.2 8.61 11.13 8.5 11.08 8.38C11.03 8.26 11 8.13 11 8C11 7.87 11.03 7.74 11.08 7.62Z"
              />
            </svg>
          </div>
          <div
            class="h-full flex-grow min-h-10 ms-2 flex flex-col box-border items-start text-inherit justify-center"
          >
            <div
              class="pl-[1px] text-small font-normal text-inherit"
            >
              You can sort your tasks by status, due date or priority. Mark tasks as completed when you finish them.
            </div>
          </div>
        </div>
        <div
          data-testid="user-task-table"
        >
          UserTaskTable Mock
        </div>
      </div>
    </section>
  </div>
</div>
`;
